src/ledger/LedgerStateTests.cpp

```diff
@@ -0,0 +1,655 @@
+// Copyright 2017 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "ledger/LedgerEntryReference.h"
+#include "ledger/LedgerHeaderReference.h"
+#include "ledger/LedgerState.h"
+#include "ledger/LedgerTestUtils.h"
+#include "lib/catch.hpp"
+#include "main/Application.h"
+#include "test/TestUtils.h"
+#include "test/test.h"
+#include <xdrpp/autocheck.h>
+
+// TODO(jonjove): Remove this header (used for LedgerEntryKey)
+#include "ledger/EntryFrame.h"
+
+using namespace stellar;
+
+LedgerEntry
+generateLedgerEntry(LedgerEntryType let)
+{
+    LedgerEntry le;
+    do
+    {
+        le = LedgerTestUtils::generateValidLedgerEntry();
+    } while (le.data.type() != let);
+    return le;
+}
+
+LedgerEntry
+generateLedgerEntryWithSameKey(LedgerEntry const& entry)
+{
+    LedgerEntryType let = entry.data.type();
+    auto le = generateLedgerEntry(let);
+    switch (let)
+    {
+    case ACCOUNT:
+        le.data.account().accountID = entry.data.account().accountID;
+        break;
+    case TRUSTLINE:
+        le.data.trustLine().accountID = entry.data.trustLine().accountID;
+        le.data.trustLine().asset = entry.data.trustLine().asset;
+        break;
+    case OFFER:
+        le.data.offer().sellerID = entry.data.offer().sellerID;
+        le.data.offer().offerID = entry.data.offer().offerID;
+        break;
+    case DATA:
+        le.data.data().accountID = entry.data.data().accountID;
+        le.data.data().dataName = entry.data.data().dataName;
+        break;
+    default:
+        abort();
+    }
+    return le;
+}
+
+TEST_CASE("LedgerState::commit", "[ledger][ledgerstate][commit]")
+{
+    for (auto val : xdr::xdr_traits<LedgerEntryType>::enum_values())
+    {
+        LedgerEntryType let = static_cast<LedgerEntryType>(val);
+        std::string letStr = xdr::xdr_traits<LedgerEntryType>::enum_name(let);
+        SECTION(letStr)
+        {
+            SECTION("[LedgerState::commit] if (mParent) { ... }")
+            {
+                VirtualClock clock;
+                Config cfg = getTestConfig();
+                cfg.INVARIANT_CHECKS = {};
+                Application::pointer app = createTestApplication(clock, cfg);
+
+                auto leParent1 = generateLedgerEntry(let);
+                auto leParent2 = generateLedgerEntryWithSameKey(leParent1);
+                auto parentKey = LedgerEntryKey(leParent1);
+                REQUIRE(parentKey == LedgerEntryKey(leParent2));
```

Comment:
is this check required? if so, maybe place it in generateLedgerEntryWithSameKey ?